in-progress/8025-publishing-tx-objects/design.md

```diff
@@ -48,27 +69,159 @@ To the DA oracle contract, the proposer will publish the following data:
-    - coinbase
-    - feeRecipient
+When the data is to be submitted as part of the same transaction as the block proposal to the L1 contract, we have two variations possible:
 
-#### DA Oracle Changes
+1. Remove the data availability oracle in its whole, and simply use the versioned KZG hashes directly
+2. Save the versioned KZG hashes in the data availability oracle
 
-Keeps track of the versioned KZG hashes that were present in a transaction.
+Solution 1 is significantly cheaper as it don't update storage. Furthermore, by requiring that the versioned KZG hashes are part of the header provided, we can show the exact KZG hashes using just the transaction and the commitment to the block hash, see `BlockLog.blockHash`.
 
 #### Rollup Contract Changes
+> There is probably some 1-off from the `count` `block_number` like values.
+
+```python
+struct State:
+  pending_proposal_count: uint64;
+  pending_slot: uint64;
+  proven_block_count: uint64;
+  proven_slot: uint64;
+
+struct ProposalLog:
+  proposal_hash: bytes32;
+  slot: uint256;
+  archive: bytes32;
+
+proposals: public(HashMap[uint256, ProposalLog])
+state: public(State);
+
+# The old `process`
+def propose(proposal, proposer_sig, attestations):
+  '''
+  Notice that no signatures are checked, all of that is optimistic 
+  '''
+  assert hash([h for h in tx.blob_versioned_hashes]) == proposal.kzgHashes # Replaces DA oracle
+  assert proposal.gas_settings.is_sane() # To be designed
+  assert proposal.slot > state.pending_slot and proposal.slot == get_current_slot()
+  assert proposal.timestamp == get_time_at_slot(proposal.slot) and proposal.timestamp <= block.timestamp
+  assert proposal.block_number == state.pending_proposal_count
+  assert proposal.inHash == INBOX.consume(proposal.block_number)
+
+  self.proposals[proposal.block_number] = ProposalLog(sha256(proposal,  proposer_sig, attestations), proposal.slot)
+  self.state.pending_proposal_count += 1
+  self.state.pending_slot = proposal.slot
+
+def challenge_proposal(proposal, proposer_sig, attestations):
+  '''
+  Implement challenges for:
+  - Bad proposer
+  - Bad attestation
+  '''
+  pass
+
+
+def submit_proof(proof, archive):
+  '''
+  Notice that this only supports sequential proofs as we always feed `self.state.proven_archive` as start state.
+  '''
+  proposal_hashes = []
+  block_number_start = self.state.proven_block_count
+  end_slot = get_last_slot_in_epoch(get_epoch_from_slot(self.proposals[block_number_start].slot))
+
+  block_number = block_number_start
+
+  for bn in range(block_number_start, block_number_start + EPOCH_LENGTH):
+    proposal_log = self.proposals[bn]
+    if proposal_log.slot <= end_slot:
+      proposal_hashes.append(proposal_log.proposal_hash)
+      block_number = bn
+    else:
+      break
+
+  assert len(proposal_hashes) > 0, 'no proposals'
+
+  prev_archive = self.proposals[self.state.proven_block_count - 1]
+  assert proof.verify(prev_archive, proposal_hashes, archive)
+
+  self.payout_fees();
```

Comment:
Currently, `total_fees` is in the `Header`, which is supplied to `submitBlockRootProof`. Seems it'll need a way to either pass a list of these `total_fees` to this function and distribute rewards for each block, or aggregate the fees to a final sum for the epoch and pass that into this function and divide the payout evenly.

e.g. if an epoch has 2 proven blocks with fees 10 and 20, the contract can either:
**Option A:** Pass the list [10, 20] and pay out, for example
TX 1: 10 -> 70% to the proposer, 20% to validators, 10% to treasury
TX 2: 20 -> same

**Option B:** Sum [10, 20] in the rollup circuits and pay out 
30 -> 70% to proposers (divided evenly, if you were proposer multiple times you get counted multiple times), 20% to validators (same), 10% to treasury

Speaking of, where/how is the contract going to get the proposer's eth address?

Reply:
BTW, my intuition tells me option A will be prohibitively expensive.

Reply:
I think that is essentially what some of Miranda's work does in submit. I just did not bother to have it in here for brevity. Should probably just extend it even if not super concise so there is less confusion.

Providing the data is not necessarily prohibitively expensive, but performing payouts for a full epoch might be, so there is the potential that it is made as a Merkle drop where the proposer needs to claim.

You have the proposer address from the proposer_sig in each proposal using ecrecover